src/components/NewDesign/Swap/Swap.tsx

```diff
@@ -0,0 +1,339 @@
+  const [tokenY, setTokenY] = React.useState<SwapToken[] | null>(null)
+  // const firstUpdate = useRef(true)
+
+  const calculateSwapOutAmount = (assetIn: SwapToken, assetFor: SwapToken, amount: string) => {
+    // TODO: solution if 0 => change to 1
+    if (priceProportion.eqn(0)) {
+      setPriceProportion(new BN(1))
+    }
+
+    const amountOut: BN =
+      swap
+        ? printBNtoBN(amount, assetIn.decimal).mul(priceProportion)
+        : printBNtoBN(amount, assetIn.decimal).div(priceProportion)
+
+    return printBN(amountOut, assetFor.decimal)
+  }
+
+  useEffect(() => {
+    updateEstimatedAmount()
+
+    if ((tokenFromIndex !== null && tokenToIndex === null)) {
+      setAmountFrom('0.000000')
+    }
+    if (tokenFromIndex !== null) {
+      const tokensY = tokens.filter((token) => {
+        if (swap) {
+          return getSwapPoolIndex(token.assetAddress, tokens[tokenToIndex ?? 0].assetAddress) !== -1
+        } else {
+          return getSwapPoolIndex(token.assetAddress, tokens[tokenFromIndex].assetAddress) !== -1
+        }
+      })
+      setTokenY(tokensY)
+    }
+  }, [tokenToIndex, tokenFromIndex])
+
+  useEffect(() => {
+    if (tokenFromIndex === null || tokenToIndex === null) {
+
+    } else {
+      const pairIndex = pools.findIndex((pool) => {
+        return (tokens[tokenFromIndex].assetAddress.toString() === pool.tokenX._bn.toString() &&
+        tokens[tokenToIndex].assetAddress.toString() === pool.tokenY._bn.toString()) ||
+       (tokens[tokenToIndex].assetAddress.toString() === pool.tokenX._bn.toString() &&
+        tokens[tokenFromIndex].assetAddress.toString() === pool.tokenY._bn.toString())
+      })
+      if (pairIndex !== -1) {
+        setPriceProportion(pools[pairIndex].sqrtPrice.v.div(new BN(10 ** PRICE_DECIMAL))
+          .mul(pools[pairIndex].sqrtPrice.v.div(new BN(10 ** PRICE_DECIMAL))))
+      }
+    }
+  }, [tokenToIndex, tokenFromIndex, swap])
+
+  useEffect(() => {
+    swap ? setTokenToIndex(tokenToIndex) : setTokenToIndex(null)
+  }, [tokenFromIndex])
+
+  useEffect(() => {
+    swap ? setTokenFromIndex(null) : setTokenFromIndex(tokenFromIndex)
+  }, [tokenToIndex])
+
+  const getSwapPoolIndex = (fromToken: PublicKey, toToken: PublicKey) => {
+    return pools.findIndex((pool) => {
+      return (pool.tokenX._bn.toString() === fromToken.toString() &&
+      pool.tokenY._bn.toString() === toToken.toString()) ||
+      (pool.tokenX._bn.toString() === toToken.toString() &&
+      pool.tokenY._bn.toString() === fromToken.toString())
+    })
+  }
+  const getIsXToY = (fromToken: PublicKey, toToken: PublicKey) => {
+    const swapPool = pools.find(
+      pool =>
+        (fromToken.toString() === pool.tokenX._bn.toString() &&
+          toToken.toString() === pool.tokenY._bn.toString()) ||
+        (fromToken.toString() === pool.tokenY._bn.toString() &&
+          toToken.toString() === pool.tokenX._bn.toString())
+    )
+
+    if (!swapPool) {
```

Comment:
this looks ugly, just use "return !!swapPool"

Reply:
done, nice case with this double ! operator :)